### DISCUSSION TITLE
Split UseMauiCommunityToolkit

### DISCUSSION THREAD

**Discussion by tranb3r** [CommunityToolkit/Maui#1631] [New Feature Discussions]
My app is not using MCT Popup, but the whole Popup code cannot be trimmed from it, because when calling `UseMauiCommunityToolkit`, the popup service is registered for DI, so the trimmer cannot get rid of it.
More generally, MCT is a collection of features more or less independant from each others, so it's probably not a good idea to use a single intialization method, if we want the trimmer to be efficient.
To fix this problem, would you consider splitting `UseMauiCommunityToolkit` by feature (or feature group)? (UseMctPopup / UseMctConverters / UseMctViews ...)?
Or maybe there is a smarter way to do workaround this issue?

**Comment by pictos** [comment]
@tranb3r, the easy fix on your side is to not call the `UseMauiCommunityToolkit` and create your extension method to register the pieces that you will use, and of course, disable the analyzer to not block your build.

**Reply by tranb3r** [reply]
Of course. This works as a quick fix.

But this is not really satisfying, for several reasons:
- I have to make sure my code is synchronized with the official method.
- I cannot call MCT `internal` code (not a problem right now, but could be in the future).
- I'm probably not the only one facing this issue (without realizing it's an issue).

**Comment by bijington** [comment]
I suspect we might need to migrate to the approach that .NET MAUI uses where we don't auto register the `PopupService` with the DI layer. Initially I wanted to make it more difficult for developers to encounter issues where they hadn't registered the `PopupService` but supporting trimming will likely provide more value here. We have the standup tomorrow so I will see if I can get this added to the agenda.

**Comment by tranb3r** [comment]
I've watched the standup. Let me share my feedback and frustration:
- This question is not just about PopupService. It's about each feature in this Toolkit. Right now MCT is a monolith because of this UseMCT method, and I do not think it's a good design. MCT is basically just a collection of things.
- We're not talking about a 5kB optimization. When enabling AOT, LVVM, multiple architectures, it can be up to 100kB (or even more), especially when only using a small set of features.
- Users read documentations. MCT is the only plugin that tells the user where it has to put the UseMyPlugin line of code. This is useless and even a pain for users who want to register their plugins in an extension of the MauiAppBuilder. Breaking this analyzer should not be a blocker when talking about improvements.
- Users want small apps. This is not something new. We've been using the linker (in a xamarin or maui app) for years. Of course it requires fine tuning, sometimes from the lib owner, sometimes from the lib consumer, but the gain can be huge. MCT will have to support it officially in net9, but right now people are already using the trimmer, and not hearing that is a mistake.
- Our app has ~200 assemblies (framework + plugins), and we've looked at each one of them to make sure the code is properly trimmed. We've discussed this issue with many plugin owners, and helped them improve their code, in order to make sure we can decrease as much as possible our app size. They've taking this issue seriously, for years, as much as we are, and things have improved a lot. So why does MCT have a different approach to this issue?
- I regret you take this kind of decisions during your standup, instead of using github to discuss more with users and contributors. We are happy to share our needs and to contribute. We can understand a decision has to be taken, and we can understand a decision, assuming there has been a debate. But I sometimes have the feeling that some issues/discussions are closed without really listening to the users. Maybe I'm wrong, but to me, it kind of defeats the purpose of having a _Community_ toolkit.

I don't want to be rude (and english not being my mother tongue does not help). I'm really thankful for the hard work (especially knowing that you're not paid by MS, as you keep repeating to everybody). I'm just sharing some frustration, and if by any chance somebody reads it, I'm happy to answer questions and help improve things as much as I can.

**Reply by bijington** [reply]
Thanks for your response. If it wasn't clear from the standup we haven't committed to anything yet. We know that we have to support trimming in .NET 9 and are committed to getting it sorted. Our conclusion at this stage is not to make any premature decisions and instead keep the discussion open. We haven't closed this discussion.

I believe we need to be pragmatic about the effort vs reward balance in implementing this. Yes developers would love their apps to be smaller but storage space that comes with devices isn't a limiting factor here.

It is great that you are willing to contribute in some way to achieving better trimmer support. As you said - this is a community toolkit.

**Reply by VladislavAntonyuk** [reply]
Treat UseMCT as UseMaui, that registers all controls and handlers, even those you don’t use. UseMCT ensures our Xaml works properly and doesn’t produce compilation errors. Also if you don’t use Popup trimmer should remove it. I agree with @bijington , Facebook is 500 mb app. But here we want to save 100 kb. Maybe you have specific needs or support old devices. As for user read documentation, it’s hard to say about all. And keep single method is much easier and simpler to understand than have 10 methods that you need to call

**Reply by tranb3r** [reply]
> Our conclusion at this stage is not to make any premature decisions and instead keep the discussion open. We haven't closed this discussion.

I'm glad you've clarified that.

Although, it really looks like you have 1000 reasons against doing it, the only good reason to do it being .net9.
So I won't insist more. I'm just adding a few comments, so you know I still disagree ;)
Anyway, thank you Shawn and Vlad for your answers. 

> storage space that comes with devices isn't a limiting factor here

It depends on the device. For old devices, it could be.
But there are other considerations. Bandwith when downloading the app. Memory usage and startup time when loading assemblies and registering stuff for nothing. Compilation time of your app, when you have several pass of optimizations after the trimmer has removed unused code.

> Facebook is 500 mb app

Users who want FB on their device do not challenge FB app size.
But when you are not FB, app size is one amongst many discrimant factors for users when deciding whether they'll install your app or not.

> Treat UseMCT as UseMaui

But MCT is not Maui. Maui is the framework, MCT is just another plugin (or collection of plugins, to insist on my point).
Also, my understanding is that Maui has a much more complex code base than MCT, and it also has a lot more functional bugs with a higher priority.
That being said, I'd be happy to get better trimming from Maui as well, but it is a separate issue.

> if you don’t use Popup trimmer should remove it

Well, PopupService is registered in UseMCT, which means the trimmer will considered it's used by the app even if it's not used in the user's code. Or maybe I'm missing something?

> keep single method is much easier and simpler to understand than have 10 methods that you need to call

We could totally have both. The single method for easy setup (this one would be in the documentation). The features methods for fine tuning (you can always put a disclaimer in the documentation/code if you consider it's risky to use them, or that it might break in the future).

**Reply by pictos** [reply]
English isn't my main language as well, and maybe I wasn't clear enough, so sorry for that. I'll try to be more descriptive here.

> Our app has ~200 assemblies (framework + plugins) ...

@tranb3r talking about your project, since you're using it as reference for this discussion. I don't know the context or history behind the decisions made for it, but when you care about size and startup performance you want to keep your project with less dependencies as possible and that looks very large dependency tree for a project that cares about it. So, I would recommend your team to work on get that number of assemblies down as much as possible. As you can see without Linker MCT has 81kb, so I'm pretty sure that our lib isn't a problem for you. For now, you can just ignore it.

Now let's talk about Linker (trimmer)

Adding support to a lib to be "Linker Friendly", like XCT is hard. When MS moved to one .NET the Linker changed its API and how it _works_ in general. But the concept is still the same... So let's go!

> When you read Linker, it's about the trimmer feature, ok?

First of all, we need to know how the Linker works. Linker is just a static analyzer of YOUR code, so in theory if we do:
https://github.com/CommunityToolkit/Maui/blob/02896d32571af05ede9df8d6b10b32db174e15f6/src/CommunityToolkit.Maui/AppBuilderExtensions.shared.cs#L23

and you don't use the concrete type or interface in your code, the Linker will see that and say "Oh no one is using this, I can remove it", if it doesn't work this way the Linker would be very poor at its job, imagine in the BCL a lot of code that's referenced internally, and it gets removed if you, the user, don't call any of those APIs. So with that in mind let's go ahead.

> Well, PopupService is registered in UseMCT, which means the trimmer will considered it's used by the app even if it's not used in the user's code. Or maybe I'm missing something?

This is not true, the Linker looks at YOUR code, not in the libs on your project. It makes decisions on what should be removed or not based on what you write. Imagine how inefficient Linker would be if it takes count looking into lib code to make decisions. So, I doubt if we register something on our builder extension that will prevent Linker to remove it.

> Users want small apps. This is not something new. We've been using the linker (in a xamarin or maui app) for years. Of course it requires fine tuning, sometimes from the lib owner, sometimes from the lib consumer, but the gain can be huge. MCT will have to support it officially in net9, but right now people are already using the trimmer, and not hearing that is a mistake.

And all of our users can have it now! We don't block anyone that wants to trim MCT, you can do it TODAY, and it works! I couldn't inspect the `.dll` because I'm on vacation and don't have all the will to do it right now, but in a small test I could confirm that you can trim our lib. So here's a small experiment

1. Create a sample app using the MAUI template
2. Add the MCT and use one control (I used the DrawingView)
3. Add `<TrimMode>full</TrimMode>` on your PropertyGroup (you can learn more of it [here](https://learn.microsoft.com/en-us/dotnet/core/deploying/trimming/trimming-options?pivots=dotnet-8-0#trimming-granularity))
4. Do a Release build and inspect the generated apk

Here's the results for NO trimming:
![image](https://github.com/CommunityToolkit/Maui/assets/20712372/35db267c-49b8-43a9-b714-5dff72d77206)

Here's the results for trimming ENABLED:
![image](https://github.com/CommunityToolkit/Maui/assets/20712372/1048b348-962f-41ec-9ab5-d9dab78e5780)

as you can see our `dll` reduces a lot in size. With that I will stress again, **we don't block anyone that wants to trim unused code from our package, we just don't work our APIs to be Linker friendly**. And that work is planned to do for .NET 9 with .NET MAUI.

**Reply by tranb3r** [reply]
I think you're wrong. Calling `UseMCT` from the app code will execute `builder.Services.AddSingleton<IPopupService, PopupService>()`, even if the app code does not use `IPopupService` and `PopupService`. So `IPopupService` and `PopupService` cannot be trimmed. This is different from a class in the BCL, that is not registered in DI or called by any static code, so that the trimmer can remove it the app code is not using it.

Here is a screenshot of what's never trimmed from MCT (new android app, do not use any MCT control, but call UseMCT):
![image](https://github.com/CommunityToolkit/Maui/assets/12516566/c8bf8060-83b2-424a-a87d-5228e82a63f0)

Anyway. We agree this is not urgent. We agree the potential gain is not huge. And there is simple workaround which consists in not calling UseMCT when the app does not use controls that need this initializer. So I guess I'll stop trying to convince you this would be a better design to split the UseMCT method...

**Reply by VladislavAntonyuk** [reply]
We may add parameters (RegisterHandlers, RegisterServices) to UseMCT. That is the simplest solution and even not a breaking change.
Parameters are enabled by default

**Reply by tranb3r** [reply]
I'm pretty sure it won't work. It won't remove code inside of a method. It has to be a different method.
And again, adding new methods is not a breaking change...

**Reply by pictos** [reply]
> I think you're wrong. Calling UseMCT from the app code will execute builder.Services.AddSingleton<IPopupService, PopupService>(), even if the app code does not use IPopupService and PopupService. So IPopupService and PopupService cannot be trimmed. 

I can be wrong, but your reason is wrong as well. The linker doesn't know what the code will do because it doesn't run the code... It's due in compilation time, not runtime... So nothing is registered in DI because there's no DI in compilation time.


@tranb3r how do you get that `dll` is from obj or bin folder? If so I think that will not show what's removed or not

**Reply by tranb3r** [reply]
> So nothing is registered in DI because there's no DI in compilation time

By "registered in DI" I meant having this kind of code in your code path: `builder.Services.AddSingleton<IPopupService, PopupService>()`
Of course the trimmer does not care about DI registration, all it sees is that your code path uses both the service interface and service class.
So the trimmer won't remove them. It's only removing the unused content (methods, properties) inside those interface/classes. Here, the constructor for PopupService is kept. And a few static stuff.

> How do you get that dll? I think that will not show what's removed or not.

Of course you have to inspect the right assembly, the one after trimming.
Two options:
- `obj\Release\net8.0-android34.0\android-arm64\linked\` folder (substitute whatever configuration/tfm you're using)
- from the `assemblies` folder in your apk. Just make sure you disable compression: `AndroidUseAssemblyStore=false` `AndroidEnableAssemblyCompression=false`

You also have to make sure to disable IL strip if you're using it: `AndroidStripILAfterAOT=false`

**Reply by tranb3r** [reply]
I also want to clarify one thing that maybe has been mis-inerpreted in my comments:
- when I say that a class is not trimmed, I mean that the class itself is kept, but its content might have been trimmed, so maybe the only thing that is left is the constructor, or some static properties.
- as opposed to a class that is totally removed by the trimmer, because the code is not referencing it _at all_.

**Reply by tranb3r** [reply]
> We may add parameters (RegisterHandlers, RegisterServices) to UseMCT

Vlad, I'm not totally sure what you have in mind.
So just to clarify:

This won't work:
```
// --- MCT code
public void UseMCT(bool useFeatureA=true, bool useFeatureB=true)
{
    if (useFeatureA) UseFeatureA();
    if (useFeatureB) UseFeatureB();
}
private void UseFeatureA() {};
private void UseFeatureB() {};
// --- User code
UseMCT(useFeatureA=true, useFeatureB=false); // UseFeatureB method will not be trimmed
```

But, as I'm suggesting, this will work:
```
// --- MCT code
public void UseMCT()
{
    UseFeatureA();
    UseFeatureB();
}
public void UseFeatureA() {};
public void UseFeatureB() {};
// --- User code
UseFeatureA(); // not calling UseFeatureB so UseFeatureB method will be trimmed
```

**Comment by CliffAgius** [comment]
Discussed again on the Monthly stand-up for [September](https://www.youtube.com/live/vNf4e5PwcTc) and the community agreed this is something that would be nice but with limited resources not something we can progress at this time.

However we are a community and more than happy to accept a PR if you feel like jumping in. 😉

**Reply by tranb3r** [reply]
@CliffAgius @brminnick 

I've listened to the stand-up and I want to clarify two things:
- I totally understand this is not high priority, this is _only_ optimization. It does not mean this discussion has to be closed without agreeing on a design for future stuff. It would be so cool if each new feature added to MCT was implemented to be compliant with trimming. But for that we have to agree on a design, which you seem reluctant to do.
- I'm not _asking_ you to implement anything, at the moment this is only a discussion, and of course I'm vocal about it, which means I'd be glad to contribute, although you're very strict about what you're accepting, so I must be sure we agree on what I'd do before starting a PR.

Here is a PR, very simple, because right now I only need `NavigationBar` in my app.
Let me know what you think about it.
https://github.com/CommunityToolkit/Maui/pull/2185
